lizard_ext/lizardns.py

```diff
@@ -0,0 +1,137 @@
+"""
+This extension counts nested control structures within a function.
+The extension is implemented with C++ in mind.
+
+The code borrows heavily from implementation of Nesting Depth extension
+originally written by Mehrdad Meh and Terry Yin.
+"""
+
+from lizard import FileInfoBuilder, FunctionInfo
+from lizard_ext.lizardnd import patch, patch_append_method
+
+
+DEFAULT_NS_THRESHOLD = 3
+
+
+class LizardExtension(object):  # pylint: disable=R0903
+
+    FUNCTION_CAPTION = ["  NS  "]
+    FUNCTION_INFO_PART = ["max_nested_structures"]
+
+    @staticmethod
+    def set_args(parser):
+        parser.add_argument(
+            "--NS",
+            help='''Threshold for the number of nested control structures.
+            The default value is %d.
+            ''' % DEFAULT_NS_THRESHOLD,
+            type=int,
+            dest="NS",
+            default=DEFAULT_NS_THRESHOLD)
+
+    def __call__(self, tokens, reader):
+        """The intent of the code is to detect control structures as entities.
+
+        The complexity arises from tracking of
+        control structures without brackets.
+        The termination of such control structures in C-like languages
+        is the next statement or control structure with a compound statement.
+
+        Moreover, control structures with two or more tokens complicates
+        the proper counting, for example, 'else if'.
+
+        In Python with meaningful indentation,
+        tracking the indentation levels becomes crucial
+        to identify boundaries of the structures.
+        The following code is not designed for Python.
+        """
+        structures = set(['if', 'else', 'foreach', 'for', 'while', 'do',
+                          'try', 'catch', 'switch'])
+
+        structure_indicator = "{"
+        structure_end = "}"
+        indent_indicator = ";"
+
+        for token in tokens:
+            if reader.context.is_within_structure():
```

Comment:
That indeed simplifies the code.
By the way, is it possible to get brace/bracket counts from the reader?